src/defrag.c

```diff
@@ -1118,75 +1096,83 @@ void activeDefragCycle(void) {
+        } defragStage;
+        defragStage defrag_stages[] = {
+            {db->keys, defragScanCallback, db},
+            {db->expires, scanCallbackCountScanned, NULL},
+            {server.pubsub_channels, defragPubsubScanCallback,
+                &(defragPubSubCtx){server.pubsub_channels, getClientPubSubChannels}},
+            {server.pubsubshard_channels, defragPubsubScanCallback,
+                &(defragPubSubCtx){server.pubsubshard_channels, getClientPubSubShardChannels}},
+        };
         do {
+            int num_stages = sizeof(defrag_stages) / sizeof(defrag_stages[0]);
+            serverAssert(defrag_stage < num_stages);
+            defragStage *current_stage = &defrag_stages[defrag_stage];
+
             /* before scanning the next bucket, see if we have big keys left from the previous bucket to scan */
             if (defragLaterStep(db, slot, endtime)) {
                 quit = 1; /* time is up, we didn't finish all the work */
                 break; /* this will exit the function and we'll continue on the next cycle */
             }
 
-            /* This array of structures holds the parameters for all defragmentation stages. */
-            defragStage defrag_stages[] = {
-                {db->keys, defragScanCallback, &(defragCtx){ db, slot }},
-                {db->expires, scanCallbackCountScanned, NULL},
-                {(current_db == 0 && slot == 0) ? server.pubsub_channels : NULL, defragPubsubScanCallback,
-                    &(defragPubSubCtx){ server.pubsub_channels, getClientPubSubChannels, slot }},
-                {server.pubsubshard_channels, defragPubsubScanCallback,
-                    &(defragPubSubCtx){ server.pubsubshard_channels, getClientPubSubShardChannels, slot }}
-            };
             if (!defrag_later_item_in_progress) {
-                /* Iterates over all defragmentation stages, starting from the last unfinished stage. */
-                int num_stages = sizeof(defrag_stages) / sizeof(defrag_stages[0]);
-                serverAssert(defrag_stage < num_stages);
-                for (int i = defrag_stage; i < num_stages; i++) {
-                    if (defrag_stage != i) break; /* previous stage not finished */
-                    if (defrag_stages[i].kvs) {
-                        defrag_cursor = kvstoreDictScanDefrag(defrag_stages[i].kvs, slot, defrag_cursor,
-                            defrag_stages[i].scanfn, &defragfns, defrag_stages[i].ctx);
-                    }
-
-                    /* Move to next stage if current stage completed */
-                    if (!defrag_cursor) defrag_stage++;
-                }
-
-                /* All stages completed */
-                if (defrag_stage == num_stages) defrag_stage = 0;
+                /* Continue defragmentation from the previous stage.
+                 * If slot is -1, it means this stage starts from the beginning. */
+                if (slot == -1) slot = kvstoreFindDictIndexByKeyIndex(current_stage->kvs, 1);
```

Comment:
maybe we should create a kvstoreGet**First**NonEmptyDictIndex instead of calling kvstoreFindDictIndexByKeyIndex(1) directly?
in kvstoreIteratorInit there's at least a comment that explains it, but creating a wrapper would make it even better.

Reply:
I'm somewhat happy with current implementation, excep for the interspersed `defragLaterItem` which is only for db.

Reply:
is that a response about kvstoreGetFirstNonEmptyDictIndex? or the other (general) question?

Reply:
do you mean that we excessively call defragLaterItem even for defrag steps that don't need it?
it doesn't cause any harm, and maybe some day other defrag steps will need it.

Reply:
> is that a response about kvstoreGetFirstNonEmptyDictIndex? or the other (general) question?

i'm working with `kvstoreGetFirstNonEmptyDictIndex`, actually i tried to make `kvstoreGetNextNonEmptyDictIndex` support slot -1, and gave up on it because it might affect the assertion that didx==0 when num_dicts is 1.

> do you mean that we excessively call defragLaterItem even for defrag steps that don't need it?

yes.

> it doesn't cause any harm, and maybe some day other defrag steps will need it.

sure, it's easy to abstract now.